[Site isolation] Snapshotting should be composited in GPUProcess and sent directly to UIProcess

shallawa opened the pull request:
#### a1a9b4a3de6f7f09d36158119cc13563d0478bc2
<pre>
[Site isolation] Snapshotting should be composited in GPUProcess and sent directly to UIProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=284427">https://bugs.webkit.org/show_bug.cgi?id=284427</a>
<a href="https://rdar.apple.com/141254954">rdar://141254954</a>

Reviewed by Matt Woodrow.

Site isolation puts a hard restriction on snapshotting. No WebContent process
is allowed to see the drawing of another remote frame drawing including the main
frame WebContent process. So the drawing of all frames will happen in GPUProcess
and the result will be sent directly to UIProcess.

The plan is to make all the remote frames draw their contents asynchronously into
separate DisplayLists in GPUProcess and then composite all these DisplayLists to
a PDFDocument or to an ShareableBitmap. This final result will sent directly to
UIProcess.

This PR is a step towards achieving this plan for site-isolation. In this PR the
main frame will be only snapshotted in GPUProcess and the result PDF SharedBuffer
will be sent to UIProcess.

The current workflow for PDF snapshotting when site-isolation = off is:

1. [WKWebView createPDFWithConfiguration] calls WebPageProxy::drawToPDF and sends
   a callback to be executed once WebProcess replies back.
2. WebPageProxy::drawToPDF sends the message WebPage::DrawToPDF to the main frame
   WebProcess.
3. WebPage::drawToPDF creates a local ImageBuffer with PDFDocument backend. Then
   it draws the main frame into this ImageBuffer.
4. Once the drawing is finished, the ImageBuffer is sunk into a SharedBuffer which
   holds the PDFDocument. The SharedBuffer is sent back to UIProcess through the
   completion handler.
5. UIProcess receives the ShareBuffer back from WebProcess and runs the callback
   which saves the returned buffer to an external file.

The new workflow for PDF snapshotting when site-isolation = on is:

1. [WKWebView createPDFWithConfiguration] calls WebPageProxy::drawCompositedToPDF()
   and sends a callback to be executed once WebProcess replies back.
2. WebPageProxy::drawCompositedToPDF() generates a new SnapshotIdentifier and adds
   the callback into the HashMap `m_pdfSnapshots` which is indexed by SnapshotIdentifier.
   Then it sends the message WebPage::DrawCompositedToPDF with SnapshotIdentifier.
3. WebPage::drawCompositedToPDF() creates a RemoteImageBufferProxy with PDFDocument
   backend. Then it draws the main frame into this ImageBuffer.
4. Once the drawing is finished, it sends RemoteRenderingBackend::DidDrawCompositedToPDF
   along with RenderingResourceIdentifier of the RemoteImageBuffer, the
   SnapshotIdentifier and the PageIdentifier
5. RemoteRenderingBackend::didDrawCompositedToPDF() gets the specified ImageBuffer
   and sinks to a PDFDocument. Then it calls GPUProcess::didDrawCompositedToPDF()
6. GPUProcess::didDrawCompositedToPDF() sends GPUProcessProxy::DidDrawCompositedToPDF
   with the SharedBuffer, the SnapshotIdentifier and the PageIdentifier.
7. GPUProcessProxy::didDrawCompositedToPDF() uses the PageIdentifier to find the
   sender WebPageProxy. Then it calls WebPageProxy::didDrawCompositedToPDF().
8. WebPageProxy::didDrawCompositedToPDF() gets the associated callback using the
   SnapshotIdentifier and runs it which saves the SharedBuffer into a file.

Drawing to PDF context in GPUProcess is done by:

1. The new backend ImageBufferRemotePDFDocumentBackend is added to act like a null
   backend for PDF snapshotting in the WebProcess.
2  RemoteRenderingBackendProxy will create an RemoteImageBufferProxy with this
   backend if renderingMode is PDFDocument.
3. The DisplayList items BeginPage and EndPage are added to record beginPage()
   and endPage() of GraphicsContext.
4. The RemoteDisplayListRecorder messages BeginPage and EndPage are added to
   record beginPage() and endPage() of GraphicsContext in GPUProcess.
5. The new message RemoteImageBuffer::SinkToPDFDocument is added which returns
   the PDF document as SharedBuffer.
6. Make RemoteImageBufferProxy implement the virtual method sinkToPDFDocument
   and make it send the sync message RemoteImageBuffer::SinkToPDFDocument.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::sinkIntoPDFDocument):
(WebCore::ImageBuffer::sinkToPDFDocument): Deleted.
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::calculateSafeBackendSize):
(WebCore::ImageBufferBackend::sinkIntoPDFDocument):
(WebCore::ImageBufferBackend::sinkToPDFDocument): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/NullImageBufferBackend.h:
* Source/WebCore/platform/graphics/SnapshotIdentifier.h: Added.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::calculateSafeBackendSize): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.cpp:
(WebCore::ImageBufferCGPDFDocumentBackend::sinkIntoPDFDocument):
(WebCore::ImageBufferCGPDFDocumentBackend::sinkToPDFDocument): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::ClipToImageBuffer::apply const):
(WebCore::DisplayList::DrawImageBuffer::apply const):
(WebCore::DisplayList::BeginPage::apply const):
(WebCore::DisplayList::BeginPage::dump const):
(WebCore::DisplayList::EndPage::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::BeginPage::BeginPage):
(WebCore::DisplayList::BeginPage::pageSize const):
(WebCore::DisplayList::EndPage::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::beginPage):
(WebCore::DisplayList::RecorderImpl::endPage):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::didDrawCompositedToPDF):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::beginPage):
(WebKit::RemoteDisplayListRecorder::endPage):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didCreateImageBuffer):
(WebKit::RemoteRenderingBackend::didDrawCompositedToPDF):
(WebKit::allocateImageBufferInternal):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Platform/IPC/ObjectIdentifierReference.serialization.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::didDrawCompositedToPDF):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawCompositedToPDF):
(WebKit::WebPageProxy::didDrawCompositedToPDF):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemotePDFDocumentBackend.cpp: Added.
(WebKit::ImageBufferRemotePDFDocumentBackend::calculateBytesPerRow):
(WebKit::ImageBufferRemotePDFDocumentBackend::calculateMemoryCost):
(WebKit::ImageBufferRemotePDFDocumentBackend::create):
(WebKit::ImageBufferRemotePDFDocumentBackend::debugDescription const):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemotePDFDocumentBackend.h: Copied from Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::beginPage):
(WebKit::RemoteDisplayListRecorderProxy::endPage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
(WebKit::RemoteRenderingBackendProxy::didDrawCompositedToPDF):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::pdfSnapshotAtSize):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawMainFrameToPDF):
(WebKit::WebPage::drawToPDF):
(WebKit::WebPage::drawCompositedToPDF):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::drawToPDFiOS):

Canonical link: <a href="https://commits.webkit.org/288113@main">https://commits.webkit.org/288113@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea5fa3dc436e411ec373d229cdc126a489a22d73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1418 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86448 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32893 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83997 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1452 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9241 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63873 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21600 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84961 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/1075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74525 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44157 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/81344 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/975 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28702 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31342 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29311 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87879 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9133 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6510 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72234 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9318 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70344 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71457 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17810 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15552 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14474 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9084 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14620 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8924 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12449 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10732 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1954,9 +1954,10 @@ - (void)createPDFWithConfiguration:(WKPDFConfiguration *)pdfConfiguration comple
     if (pdfConfiguration && !CGRectIsNull(pdfConfiguration.rect))
         floatRect = WebCore::FloatRect(pdfConfiguration.rect);
 
+    auto drawToPDFFunction = _page->preferences().siteIsolationEnabled() ? &WebKit::WebPageProxy::drawCompositedToPDF : &WebKit::WebPageProxy::drawToPDF;
```
Should we consider a separate setting for this, and enable it by default (or if site isolation is enabled).

The idea would be to get people using the new path and get it tested before site-isolation ships (with the option to disable it quickly if things go wrong).

shallawa replied:
I agree that testing this part should be separate from the site isolation feature. I will socialize this topic to get an agreement and see what the proper name for this feature should be.

shallawa replied:
I will submit this patch using the `siteIsolationEnabled` setting for now. And I may submit another patch for this composited snapshotting later.

shallawa replied:
I will remove the iOS changes also because I realized it is more complicated than macOS and it may require snapshotting one PDF page from the web page. Or it may require all the PDF pages. And it may work on a background thread or the main thread. This makes the spans hotting work sometimes synchronously.

shallawa committed:
[Site isolation] Snapshotting should be composited in GPUProcess and sent directly to UIProcess
https://bugs.webkit.org/show_bug.cgi?id=284427
rdar://141254954

Reviewed by Matt Woodrow.

Site isolation puts a hard restriction on snapshotting. No WebContent process
is allowed to see the drawing of another remote frame drawing including the main
frame WebContent process. So the drawing of all frames will happen in GPUProcess
and the result will be sent directly to UIProcess.

The plan is to make all the remote frames draw their contents asynchronously into
separate DisplayLists in GPUProcess and then composite all these DisplayLists to
a PDFDocument or to an ShareableBitmap. This final result will sent directly to
UIProcess.

This PR is a step towards achieving this plan for site-isolation. In this PR the
main frame will be only snapshotted in GPUProcess and the result PDF SharedBuffer
will be sent to UIProcess.

The current workflow for PDF snapshotting when site-isolation = off is:

1. [WKWebView createPDFWithConfiguration] calls WebPageProxy::drawToPDF and sends
   a callback to be executed once WebProcess replies back.
2. WebPageProxy::drawToPDF sends the message WebPage::DrawToPDF to the main frame
   WebProcess.
3. WebPage::drawToPDF creates a local ImageBuffer with PDFDocument backend. Then
   it draws the main frame into this ImageBuffer.
4. Once the drawing is finished, the ImageBuffer is sunk into a SharedBuffer which
   holds the PDFDocument. The SharedBuffer is sent back to UIProcess through the
   completion handler.
5. UIProcess receives the ShareBuffer back from WebProcess and runs the callback
   which saves the returned buffer to an external file.

The new workflow for PDF snapshotting when site-isolation = on is:

1. [WKWebView createPDFWithConfiguration] calls WebPageProxy::drawCompositedToPDF()
   and sends a callback to be executed once WebProcess replies back.
2. WebPageProxy::drawCompositedToPDF() generates a new SnapshotIdentifier and adds
   the callback into the HashMap `m_pdfSnapshots` which is indexed by SnapshotIdentifier.
   Then it sends the message WebPage::DrawCompositedToPDF with SnapshotIdentifier.
3. WebPage::drawCompositedToPDF() creates a RemoteImageBufferProxy with PDFDocument
   backend. Then it draws the main frame into this ImageBuffer.
4. Once the drawing is finished, it sends RemoteRenderingBackend::DidDrawCompositedToPDF
   along with RenderingResourceIdentifier of the RemoteImageBuffer, the
   SnapshotIdentifier and the PageIdentifier
5. RemoteRenderingBackend::didDrawCompositedToPDF() gets the specified ImageBuffer
   and sinks to a PDFDocument. Then it calls GPUProcess::didDrawCompositedToPDF()
6. GPUProcess::didDrawCompositedToPDF() sends GPUProcessProxy::DidDrawCompositedToPDF
   with the SharedBuffer, the SnapshotIdentifier and the PageIdentifier.
7. GPUProcessProxy::didDrawCompositedToPDF() uses the PageIdentifier to find the
   sender WebPageProxy. Then it calls WebPageProxy::didDrawCompositedToPDF().
8. WebPageProxy::didDrawCompositedToPDF() gets the associated callback using the
   SnapshotIdentifier and runs it which saves the SharedBuffer into a file.

Drawing to PDF context in GPUProcess is done by:

1. The new backend ImageBufferRemotePDFDocumentBackend is added to act like a null
   backend for PDF snapshotting in the WebProcess.
2  RemoteRenderingBackendProxy will create an RemoteImageBufferProxy with this
   backend if renderingMode is PDFDocument.
3. The DisplayList items BeginPage and EndPage are added to record beginPage()
   and endPage() of GraphicsContext.
4. The RemoteDisplayListRecorder messages BeginPage and EndPage are added to
   record beginPage() and endPage() of GraphicsContext in GPUProcess.
5. The new message RemoteImageBuffer::SinkToPDFDocument is added which returns
   the PDF document as SharedBuffer.
6. Make RemoteImageBufferProxy implement the virtual method sinkToPDFDocument
   and make it send the sync message RemoteImageBuffer::SinkToPDFDocument.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::sinkIntoPDFDocument):
(WebCore::ImageBuffer::sinkToPDFDocument): Deleted.
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::calculateSafeBackendSize):
(WebCore::ImageBufferBackend::sinkIntoPDFDocument):
(WebCore::ImageBufferBackend::sinkToPDFDocument): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/NullImageBufferBackend.h:
* Source/WebCore/platform/graphics/SnapshotIdentifier.h: Added.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::calculateSafeBackendSize): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.cpp:
(WebCore::ImageBufferCGPDFDocumentBackend::sinkIntoPDFDocument):
(WebCore::ImageBufferCGPDFDocumentBackend::sinkToPDFDocument): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::ClipToImageBuffer::apply const):
(WebCore::DisplayList::DrawImageBuffer::apply const):
(WebCore::DisplayList::BeginPage::apply const):
(WebCore::DisplayList::BeginPage::dump const):
(WebCore::DisplayList::EndPage::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::BeginPage::BeginPage):
(WebCore::DisplayList::BeginPage::pageSize const):
(WebCore::DisplayList::EndPage::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::beginPage):
(WebCore::DisplayList::RecorderImpl::endPage):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::didDrawCompositedToPDF):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::beginPage):
(WebKit::RemoteDisplayListRecorder::endPage):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didCreateImageBuffer):
(WebKit::RemoteRenderingBackend::didDrawCompositedToPDF):
(WebKit::allocateImageBufferInternal):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Platform/IPC/ObjectIdentifierReference.serialization.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::didDrawCompositedToPDF):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawCompositedToPDF):
(WebKit::WebPageProxy::didDrawCompositedToPDF):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemotePDFDocumentBackend.cpp: Added.
(WebKit::ImageBufferRemotePDFDocumentBackend::calculateBytesPerRow):
(WebKit::ImageBufferRemotePDFDocumentBackend::calculateMemoryCost):
(WebKit::ImageBufferRemotePDFDocumentBackend::create):
(WebKit::ImageBufferRemotePDFDocumentBackend::debugDescription const):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemotePDFDocumentBackend.h: Copied from Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::beginPage):
(WebKit::RemoteDisplayListRecorderProxy::endPage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
(WebKit::RemoteRenderingBackendProxy::didDrawCompositedToPDF):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::pdfSnapshotAtSize):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawMainFrameToPDF):
(WebKit::WebPage::drawToPDF):
(WebKit::WebPage::drawCompositedToPDF):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::drawToPDFiOS):

Canonical link: https://commits.webkit.org/288113@main
